tests/tensorbox/test_functions.py

```diff
@@ -0,0 +1,519 @@
+
+    assert fn.allequal(res, t1)
+    assert onp.asarray(res).dtype.type is onp.asarray(t2).dtype.type
+
+
+class TestConvertLike:
+    """tests for the convert like function"""
+
+    @pytest.mark.parametrize("t1,t2", list(itertools.combinations(test_data, r=2)))
+    def test_convert_tensor_like(self, t1, t2):
+        """Test that converting t1 like t2 results in t1 being cast to the same tensor type as t2"""
+        res = fn.convert_like(t1, t2)
+
+        # if tensorflow or pytorch, extract view of underlying data
+        if hasattr(res, "numpy"):
+            res = res.numpy()
+
+        if hasattr(t2, "numpy"):
+            t2 = t2.numpy()
+
+        assert fn.allequal(res, t1)
+        assert isinstance(res, np.ndarray if isinstance(t2, (list, tuple)) else t2.__class__)
+
+    @pytest.mark.parametrize("t_like", [np.array([1]), tf.constant([1]), torch.tensor([1])])
+    def test_convert_scalar(self, t_like):
+        """Test that a python scalar is converted to a scalar tensor"""
+        res = fn.convert_like(5, t_like)
+        assert isinstance(res, t_like.__class__)
+        assert res.ndim == 0
+        assert fn.allequal(res, [5])
+
+
+# the following test data is of the form
+# [original shape, axis to expand, new shape]
+expand_dims_test_data = [
+    [tuple(), 0, (1,)],
+    [(3,), 0, (1, 3)],
+    [(3,), 1, (3, 1)],
+    [(2, 2), 0, (1, 2, 2)],
+    [(2, 2), 1, (2, 1, 2)],
+    [(2, 2), 2, (2, 2, 1)],
+]
+
+
+@pytest.mark.parametrize("shape,axis,new_shape", expand_dims_test_data)
+class TestExpandDims:
+    """Tests for the expand_dims function"""
+
+    def test_expand_dims_sequence(self, shape, axis, new_shape):
+        """Test that the expand dimensions works correctly
+        when given a sequence"""
+        if not shape:
+            pytest.skip("Cannot expand the dimensions of a Python scalar!")
+
+        t1 = np.empty(shape).tolist()
+        t2 = fn.expand_dims(t1, axis=axis)
+        assert t2.shape == new_shape
+
+    def test_expand_dims_array(self, shape, axis, new_shape):
+        """Test that the expand dimensions works correctly
+        when given an array"""
+        t1 = np.empty(shape)
+        t2 = fn.expand_dims(t1, axis=axis)
+        assert t2.shape == new_shape
+        assert isinstance(t2, np.ndarray)
+
+    def test_expand_dims_torch(self, shape, axis, new_shape):
+        """Test that the expand dimensions works correctly
+        when given a torch tensor"""
+        t1 = torch.empty(shape)
+        t2 = fn.expand_dims(t1, axis=axis)
+        assert t2.shape == new_shape
+        assert isinstance(t2, torch.Tensor)
+
+    def test_expand_dims_tf(self, shape, axis, new_shape):
+        """Test that the expand dimensions works correctly
+        when given a TF tensor"""
+        t1 = tf.ones(shape)
+        t2 = fn.expand_dims(t1, axis=axis)
+        assert t2.shape == new_shape
+        assert isinstance(t2, tf.Tensor)
+
+
+interface_test_data = [
+    [(1, 2, 3), "numpy"],

```

Comment:
what happens if we feed scalars? this might happen a lot actually!

Reply:
Ah, feeding in scalars will raise an error currently, since the `TensorBox` won't be able to deduce which framework to dispatch to 🤔